### PR TITLE
feat: validate and skip foreign key violation records in child (RDCC-2082)

### PR DESCRIPTION
**Pull request by abhijitdiwan**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/RDCC-2082

### Change description ###
feat: validate and skip foreign key violation records in child (RDCC-2082)

**Does this PR introduce a breaking change?** (check one with "x")
No

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/juddata/camel/processor/ICustomValidationProcessor.java [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+                    filteredRecord = filteredChildren.removeIf(filterInvalidUserProfAppointment ->
+                        ((JudicialOfficeAuthorisation) filterInvalidUserProfAppointment).getElinksId()
+                            .equalsIgnoreCase(invalidRecords.getElinksId()));
+                }
+                if (filteredRecord) {
+                    invalidElinks.add(invalidRecords.getElinksId());
+                }
+            });
+
+            if (isFalse(isEmpty(invalidElinks))) {
```
Comment (shreedhar-hmcts) [inline src/main/java/uk/gov/hmcts/reform/juddata/camel/processor/ICustomValidationProcessor.java:59]: I think we can use apache isNotEmpty() method here

File: src/main/java/uk/gov/hmcts/reform/juddata/camel/processor/ICustomValidationProcessor.java [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+            invalidJudicialUserProfileRecords.forEach(invalidRecords -> {
+                //Remove invalid appointment for user profile and add to invalidElinks List
+                boolean filteredRecord;
+                if (((Class) mySuperclass).getCanonicalName().equals(JudicialOfficeAppointment
+                    .class.getCanonicalName())) {
+                    filteredRecord = filteredChildren.removeIf(filterInvalidUserProfAppointment ->
+                        ((JudicialOfficeAppointment) filterInvalidUserProfAppointment).getElinksId()
+                            .equalsIgnoreCase(invalidRecords.getElinksId()));
+                } else {
+                    filteredRecord = filteredChildren.removeIf(filterInvalidUserProfAppointment ->
```
Comment (shreedhar-hmcts) [inline src/main/java/uk/gov/hmcts/reform/juddata/camel/processor/ICustomValidationProcessor.java:50]: rename filterInvalidUserProfAppointment to something like authorisation for better readability

File: src/main/java/uk/gov/hmcts/reform/juddata/camel/processor/ICustomValidationProcessor.java [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+
+            invalidJudicialUserProfileRecords.forEach(invalidRecords -> {
+                //Remove invalid appointment for user profile and add to invalidElinks List
+                boolean filteredRecord;
+                if (((Class) mySuperclass).getCanonicalName().equals(JudicialOfficeAppointment
+                    .class.getCanonicalName())) {
+                    filteredRecord = filteredChildren.removeIf(filterInvalidUserProfAppointment ->
+                        ((JudicialOfficeAppointment) filterInvalidUserProfAppointment).getElinksId()
+                            .equalsIgnoreCase(invalidRecords.getElinksId()));
+                } else {
```
Comment (shreedhar-hmcts) [inline src/main/java/uk/gov/hmcts/reform/juddata/camel/processor/ICustomValidationProcessor.java:49]: require elseIf and check if record is of authorisation? better for future if any new child is added ?

File: src/main/java/uk/gov/hmcts/reform/juddata/camel/processor/ICustomValidationProcessor.java (hunk 1 above)
Comment (abhijitdiwan) [reply src/main/java/uk/gov/hmcts/reform/juddata/camel/processor/ICustomValidationProcessor.java:59]: we have not using https://commons.apache.org/proper/commons-collections/apidocs/org/apache/commons/collections4/CollectionUtils.html this jar hence avoided to put here. don't know adding on more jar for sake of checking this good or bad. let me know what you think?

File: src/main/java/uk/gov/hmcts/reform/juddata/camel/processor/ICustomValidationProcessor.java [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+        if (isFalse(isEmpty(missingForeignKeyRecords))) {
+            if (((Class) mySuperclass).getCanonicalName().equals(JudicialOfficeAppointment
+                .class.getCanonicalName())) {
+                //Auditing JSR skipped rows of user profile for Appointment
+                jsrValidatorInitializer.auditJsrExceptions(missingForeignKeyRecords.stream()
+                        .map(s -> ((JudicialOfficeAppointment) s).getElinksId()).collect(toList()),
+                    fieldInError, errorMessage, exchange);
+            } else {
+                jsrValidatorInitializer.auditJsrExceptions(missingForeignKeyRecords.stream()
+                        .map(s -> ((JudicialOfficeAuthorisation) s).getElinksId()).collect(toList()),
```
Comment (shreedhar-hmcts) [inline src/main/java/uk/gov/hmcts/reform/juddata/camel/processor/ICustomValidationProcessor.java:90]: same comments here also as above

File: src/main/java/uk/gov/hmcts/reform/juddata/camel/processor/ICustomValidationProcessor.java (hunk 3 above)
Comment (abhijitdiwan) [reply src/main/java/uk/gov/hmcts/reform/juddata/camel/processor/ICustomValidationProcessor.java:49]: I think is fine for now...but lest do it :)

File: src/test/java/uk/gov/hmcts/reform/juddata/camel/helper/JrdTestSupport.java [hunk 5]
```diff
@@ -16,16 +9,28 @@
...
 
     }
 
-    public static JudicialUserProfile createJudicialUserProfileMock(Date currentDate, LocalDateTime dateTime) {
+    public static JudicialUserProfile createJudicialUserProfileMock(Date currentDate, LocalDateTime dateTime,
+                                                                    String elinksId) {
 
         JudicialUserProfile judicialUserProfileMock = new JudicialUserProfile();
-        judicialUserProfileMock.setElinksId("elinksid_1");
+        judicialUserProfileMock.setElinksId(elinksId); //"elinksid_1");
```
Comment (shreedhar-hmcts) [inline src/test/java/uk/gov/hmcts/reform/juddata/camel/helper/JrdTestSupport.java:33]: commented can be removed ?

File: src/main/java/uk/gov/hmcts/reform/juddata/camel/processor/ICustomValidationProcessor.java (hunk 1 above)
Comment (abhijitdiwan) [reply src/main/java/uk/gov/hmcts/reform/juddata/camel/processor/ICustomValidationProcessor.java:59]: need to use camel utility as we are not using https://commons.apache.org/proper/commons-collections/apidocs/org/apache/commons/collections4/CollectionUtils.html this jar :)

**Review by shreedhar-hmcts — APPROVED** [review]

**Review by shreedhar-hmcts — APPROVED** [review]

**Review by pksaxena01 — APPROVED** [review]
Approved

File: src/main/java/uk/gov/hmcts/reform/juddata/camel/processor/JudicialUserProfileProcessor.java [hunk 6]
```diff
@@ -44,6 +57,25 @@ public void process(Exchange exchange) {
...
+        //Get Elink ids from current load
+        validElinksInUserProfile = filteredJudicialUserProfiles.stream()
+            .map(userProfile -> userProfile.getElinksId()).collect(toSet());
+
+        //Get Elinks from previous loads
+        validElinksInUserProfile.addAll(loadElinksId());
+
+        filteredJudicialUserProfiles.stream()
+            .map(userProfile -> userProfile.getElinksId()).collect(toSet());
+
```
Comment (shashank-rastogi) [inline src/main/java/uk/gov/hmcts/reform/juddata/camel/processor/JudicialUserProfileProcessor.java:69]: why is it done this way? why cant we get the elinks id from db only?

Comment (shashank-rastogi) [reply src/main/java/uk/gov/hmcts/reform/juddata/camel/processor/JudicialUserProfileProcessor.java:69]: code at line 61 and 67 is repeated and can be avoided if u fetch it from db directly?

**Pull request merged**